### PR TITLE
[LINALG] Fix: Incorrect linalg lowering for aten.convolution_transpose with negative effective padding

### PR DESCRIPTION
**Pull request by hariprasadravi**
### **The Bug**

The `torch-to-linalg` lowering for `aten.convolution` (with `transposed=true`) incorrectly handles cases where the effective padding is negative. The logic for this is contained in `createTransposedInputPadding`.

The original implementation had two critical flaws:

**Incorrect Math**: The logic block for negative padding (if (anyDimensionPaddingIsNegative)) attempted to "pre-crop" the input tensor before un-striding. The math used to calculate these slice offsets and sizes was incorrect, resulting in `tensor.extract_slice` operations with out-of-bounds offsets and negative sizes, causing the compiler to fail.

**Failed "Mixed-Mode**" **Logic**: The code was built on an "all-or-nothing" assumption. It failed to handle "mixed-mode" padding, where one spatial dimension required padding (positive offset) while another required cropping (negative offset). It would enter the negative padding path and apply cropping logic to all dimensions, leading to out-of-bounds errors when it tried to crop a dimension that should have been padded.

### **The Fix**
This patch refactors the logic into two clean, robust paths:

**All-Padding Path (else block):**

Trigger: All spatial dimensions have an effective padding offset >= 0.

Action: Retains the original, efficient "fast path." It uses a single `tensor.insert_slice` to perform both un-striding (with strides) and padding (with offsets) in one operation.

**Safe Path (if (anyDimensionPaddingIsNegative) block):**

Trigger: At least one spatial dimension has a negative effective padding offset.

Action: This path is now a unified, robust 3-step process that correctly handles both all-crop and mixed-mode scenarios:

Create "Super-Tensor": It computes a maxSizes tensor, which is the "union" of the padded and un-strided sizes (i.e., max(innerSize, outerSize) for each dimension).

Pad & Un-stride: It performs a single `tensor.insert_slice` of the original input into this maxSizes tensor. This one operation correctly applies all positive padding (via insertSliceOffsets) and un-striding (via strideIndexValues).

Crop: It performs a final `tensor.extract_slice` to crop the maxSizes tensor down to the final outerSizes. This correctly applies all negative padding (via extractSliceOffsets).

This new logic resolves all known failure cases and is validated by the new TransposedConv{1,2,3}dNegativePadding test cases, which specifically target this functionality.

### CODE REVIEW

**Comment by hariprasadravi** [comment]
@sahas3 and @ivangarcia44  Please help review and add additional reviewers.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py [hunk 1]
```diff
@@ -2052,7 +2116,7 @@ def forward(self, inputVec, weight, bias):
             inputVec,
             weight,
             bias=bias,
-            stride=[1, 1, 1],
+            stride=[4, 4, 4],
```
Comment (ivangarcia44) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py]: If it does not hurt, you could use different numbers for the strides. Sometimes using non-uniform values uncover issues.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py [hunk 2]
```diff
@@ -1988,7 +1988,7 @@ def forward(self, inputVec, weight, bias):
             inputVec,
             weight,
             bias=bias,
-            stride=[1],
+            stride=[4],
```
Comment (ivangarcia44) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py:1991]: Is it necessary to change the stride to 4? Is there a test with unitary stride?

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py [hunk 3]
```diff
@@ -2034,6 +2066,38 @@ def TransposedConv2dNegativePadding_basic(module, tu: TestUtils):
     module.forward(tu.rand(1, 1, 4, 7), tu.rand(1, 2, 3, 3), tu.rand(2))
 
 
+class TransposedConv2dPositiveAndNegativePadding(torch.nn.Module):
```
Comment (ivangarcia44) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py:2103]: I would add a transposed Conv3D test with mixed negative/positive padding.

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 4]
```diff
@@ -1551,7 +1551,6 @@ Value ConvertAtenConvolutionOp::createTransposedInputPadding(
   SmallVector<Value> insertSliceOffsets{c0, c0};
 
   SmallVector<Value> inputSizes = getTensorSizes(rewriter, loc, input);
-  SmallVector<Value> sliceSizes{inputSizes[0], inputSizes[1]};
```
Comment (ivangarcia44) [inline lib/Conversion/TorchToLinalg/Linear.cpp:1554]: If you think it would help, you could add a small description at the top of what the function does (now that you have it fresh in your mind) with small pseudocode examples. If it applies, you could also add algebraic formulas for the padding since these are usually clearer than MLIR's code construction API.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py (hunk 2 above)
Comment (hariprasadravi) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py:1991]: If `stride` is 1,  the old logic would  create a `tensor.extract_slice` that would look like `tensor.extract_slice %input[..., 1] [..., 5] [..., 1]`
The verifier sees that the last element accessed is `offset + (size-1)*stride` which is `1+ (5-1)*1 == 5` and this will be within the bounds as input size is 7. 

If the stride was 4, this would be  `tensor.extract_slice %input[..., 1] [..., 5] [..., 4]` and we would be out of bounds.
We need stride to be larger than 1 to amplify the issue and lock down the fix.

That said, you're right that we need a stride == 1 testpoint. I will add it.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py (hunk 3 above)
Comment (hariprasadravi) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py:2103]: Instead of adding another test, I have modified this test to do padding in height direction and cropping in width direction

```
class TransposedConv3dNegativePadding(torch.nn.Module):
    def __init__(self):
        super().__init__()

    @export
    @annotate_args(
        [
            None,
            ([4, 1, 8, 13, 17], torch.float32, True),
            ([1, 1, 3, 7, 3], torch.float32, True),
            ([1], torch.float32, True),
        ]
    )
    def forward(self, inputVec, weight, bias):
        return torch.ops.aten.convolution(
            inputVec,
            weight,
            bias=bias,
            stride=[4, 5, 3],
            padding=[2, 1, 3],
            dilation=[1, 2, 1],
            transposed=True,
            output_padding=[0, 0, 0],
            groups=1,
        )
  ```
  
Offset calculation = `offset = (kernel_size - 1) * dilation - padding` 

Height (Dimension 1):

Kernel: 7

Dilation: 2

Padding: 1

offset_height = (7 - 1) * 2 - 1 = (6 * 2) - 1 = 11

Result: +11 (Positive Padding)

Width (Dimension 2):

Kernel: 3

Dilation: 1

Padding: 3

offset_width = (3 - 1) * 1 - 3 = (2 * 1) - 3 = -1

Result: -1 (Negative Padding / Cropping)

**Comment by ivangarcia44** [comment]
The changes LGTM. No more comments on my side. Thank you for adding the description. It is easy to see now in the formula why a large stride is going to make a negative padding.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py [hunk 5]
```diff
@@ -2034,6 +2098,38 @@ def TransposedConv2dNegativePadding_basic(module, tu: TestUtils):
...
+    def __init__(self):
+        super().__init__()
+
+    @export
+    @annotate_args(
+        [
+            None,
+            ([1, 1, 4, 7], torch.float32, True),
```
Comment (sahas3) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py:2111]: Can one of the tests specify dynamic dimensions to ensure that is handled correctly too? From the code nothing seems tied with static dims, so I think dynamic dims will just work?

Comment (hariprasadravi) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py:2111]: Yes I have modified one testpoint to test dynamic spatial dims.

Comment (hariprasadravi) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py:2111]: It seems that the ONNX backend does not like dynamic dims. I have filtered the test for ONNX
